src/lcmtype.jl

```diff
@@ -13,59 +13,141 @@ Subtypes must be `mutable struct`s and may use the following field types:
+function sizefields end
 
 """
-fingerprint(::Type{T}) where T<:LCMType
+    fingerprint(::Type{T}) where T<:LCMType
 
 Return the fingerprint of LCM type `T` as an `SVector{8, UInt8}`.
 """
-fingerprint(::Type{T}) where {T<:LCMType} = error("fingerprint method not defined for LCMType $T.")
+function fingerprint end
 
 # Types that are encoded in network byte order, as dictated by the LCM type specification.
-const NETWORK_BYTE_ORDER_TYPES = Union{Int8, Int16, Int32, Int64, Float32, Float64, UInt8}
+const NetworkByteOrderEncoded = Union{Int8, Int16, Int32, Int64, Float32, Float64, UInt8}
+
+# Julia types that correspond to LCM primitive types
+const LCMPrimitive = Union{Int8, Int16, Int32, Int64, Float32, Float64, String, Bool, UInt8}
 
 # Default values for all of the possible field types of an LCM type:
-default_value(::Type{Bool}) = false
-default_value(::Type{T}) where {T<:NETWORK_BYTE_ORDER_TYPES} = zero(T)
-default_value(::Type{String}) = ""
-default_value(::Type{T}) where {T<:Vector} = T()
-default_value(::Type{SV}) where {N, T, SV<:StaticVector{N, T}} = SV(ntuple(i -> default_value(T), Val(N)))
-default_value(::Type{T}) where {T<:LCMType} = T()
+defaultval(::Type{Bool}) = false
+defaultval(::Type{T}) where {T<:NetworkByteOrderEncoded} = zero(T)
+defaultval(::Type{String}) = ""
+defaultval(::Type{T}) where {T<:Vector} = T()
+defaultval(::Type{SV}) where {N, T, SV<:StaticVector{N, T}} = SV(ntuple(i -> defaultval(T), Val(N)))
+defaultval(::Type{T}) where {T<:LCMType} = T()
 
 # Generated default constructor for LCMType subtypes
 @generated function (::Type{T})() where T<:LCMType
-    constructor_arg_exprs = [:(default_value(fieldtype(T, $i))) for i = 1 : fieldcount(T)]
+    constructor_arg_exprs = [:(defaultval(fieldtype(T, $i))) for i = 1 : fieldcount(T)]
     :(T($(constructor_arg_exprs...)))
 end
 
-# Fingerprint check
+# Field dimension information
+@enum DimensionMode LCM_CONST LCM_VAR
+struct LCMDimension{S <: Union{Symbol, Int}}
+    size::S # either a field name, or the size of a statically-sized field
+end
+dimensionmode(::Type{LCMDimension{Symbol}}) = LCM_VAR
+dimensionmode(::Type{LCMDimension{Int}}) = LCM_CONST
+
+# Hash computation
+# Port of https://github.com/ZeroCM/zcm/blob/e9c7bfc401ea15aa64291507da37d1163e5506c0/gen/ZCMGen.cpp#L73-L94
+# NOTE: actual LCM implementation: https://github.com/lcm-proj/lcm/blob/992959adfbda78a13a858a514636e7929f27ed16/lcmgen/lcmgen.c#L114-L132
+# uses an int64_t for v and thus relies on undefined behavior; see https://stackoverflow.com/a/4009922.
+function sign_extended_right_shift(val::UInt64, nshift::UInt64)
+    (val >> 63) == 0 ? val >> nshift :
+        (val >> nshift) | ~((1 << (UInt64(64) - nshift)) - UInt64(1))
+end
+
+hashupdate(v::UInt64, c::Cchar) = ((v << 8) ⊻ sign_extended_right_shift(v, UInt64(55))) + c
+hashupdate(v::UInt64, c::Char) = hashupdate(v, Cchar(c))
+hashupdate(v::UInt64, n::Integer) = hashupdate(v, Cchar(n))
+
+function hashupdate(v::UInt64, s::AbstractString)
```

Comment:
Technically, this will produce different results than the C implementation if your field names contain non-ASCII characters, because the C implementation just blindly hashes every byte, while this version will try to call `Cchar` on each character (which will throw InexactError() for non-ASCII). 

However, constructing a message type that causes this failure would probably break every other LCM tool too (it would almost certainly cause Python 2 to explode), so I'm not too worried.

Reply:
Yeah, not too worried about this case. If anything, failing fast here sounds like the right thing to do.